breach-logos.js

```diff
@@ -0,0 +1,86 @@
+"use strict";
+
+const got = require("got");
+const fs = require("fs");
+const { promisify } = require("util");
+const writeFile = promisify(fs.writeFile);
+
+/* eslint-disable node/no-unpublished-require */
+const imagemin = require("imagemin");
+const pngQuant = require("imagemin-pngquant");
+/* eslint-enable node/no-unpublished-require */
+
+const mozlog = require("./log");
+const log = mozlog("breach-logos");
+
+
+const BreachLogos = {
+
+  remoteDir: "https://haveibeenpwned.com/Content/Images/PwnedLogos/",
+  localDir: "./public/img/logos/",
+
+  async minifyLogoData(logoData) {
+    return await imagemin.buffer(logoData, {
+      plugins: [pngQuant({quality: 10})],
+    });
+  },
+
+  async downloadLogo(fileName) {
+    const remoteSrc = `${this.remoteDir}${fileName}`;
+    const localSrc = `${this.localDir}${fileName}`;
+    let fileIsBig = false;
+    let logoData = [];
+
+    const stream = await got.stream(remoteSrc, {
+      retries: 3,
+      timeout: {
+        connect: 1000,
+        socket: 1000,
+        request: 5000,
+      },
+    });
+
+    stream.on("data", incomingData => logoData.push(incomingData));
+    stream.on("response", res => {
+
+        // Check response and error out on certain conditions.
+        if (res.headers["content-type"] !== "image/png") { /* eslint-disable quotes */
+          throw new Error(`response.headers["content-type"] !== "image/png"`); /* eslint-enable quotes */
+        }
+        if (res.statusCode !== 200) {
+          throw new Error(`Status: ${res.statusCode}`);
+        }
+        if (parseInt(res.headers["content-length"]) > 1000000) {
```

Comment:
This number was a bit arbitrary. Not sure what we think a reasonable cut off figure is.

Reply:
Is this 1MB? Do we know the average or median HIBP image size?

Reply:
Anecdotally, they're less than 10kb. There's really no reason we should expect to see an image this big come floating across but just need to decide where to draw the line in the event of an unexpected-huge-file-occurence.

Reply:
Sounds good to me!

Reply:
I actually looked into this once briefly.

So far our worst offending image I've found [from HIBP] was Parapa.png, at 231 KB:

```
464 -rw-r--r--    1 pdehaan  staff   231K 31 Dec 23:46 Parapa.png
```

But, if I view the image in the repo, it's a lovely optimized 38 KB;
```
 80 -rw-r--r--    1 pdehaan  staff    38K 31 Dec 23:48 Parapa.png
```

So I'm not sure what the best approach is... If the image is considered "large" at 500 KB, but it's almost arbitrary and we would really only care about the file size of the _optimized_ logo, and maybe error out if the minified file size is over 100 KB.

Reply:
I distinctly remembered the panic level exclamations associated with the soul-crushing Parapa logo and used it to test this set up! Pulling the existing logo from HIBP and pushing it through png-quant resulted locally in an a 37kb end-product.